# Name: Aniruddh Kulkarni 
# Roll no: I081          
# Stream: CS (AI)      
# Division: I                         
# Semester: 5th Semester                                           
# Batch: I-3            
# Subject: NLP
# Assignment-7

In [1]:
!pip install scikit-learn==0.21.3
!pip install wget==3.2
!pip install gensim==3.6.0
!pip install psutil==5.4.8
!pip install spacy==2.2.4

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 12.2/12.2 MB 20.8 MB/s eta 0:00:0000:0100:01
  Preparing metadata (setup.py) ... done
  error: subprocess-exited-with-error
  
  × python setup.py bdist_wheel did not run successfully.
  │ exit code: 1
  ╰─> [29 lines of output]
      Partial import of sklearn during the build process.
      Traceback (most recent call last):
        File "<string>", line 2, in <module>
        File "<pip-setuptools-caller>", line 34, in <module>
        File "/private/var/folders/w7/x95pzq4s5s1f44mr9jmxg5_m0000gn/T/pip-install-g06xu9ab/scikit-learn_b4abe4fa087e4f11bd4f8b60ab5b8919/setup.py", line 290, in <module>
          setup_package()
        File "/private/var/folders/w7/x95pzq4s5s1f44mr9jmxg5_m0000gn/T/pip-install-g06xu9ab/scikit-learn_b4abe4fa087e4f11bd4f8b60ab5b8919/setup.py", line 286, in setup_package
          setup(**metadata)
        File "/Users/pushpakulkarni/miniconda3/lib/python3.10/site-packages/numpy/distutils/core.py", line 135, in setu

  Preparing metadata (setup.py) ... done
  Created wheel for wget: filename=wget-3.2-py3-none-any.whl size=9656 sha256=a947fb2429af86325ae3f0d8df66c3777c838122b0872ace4d4f364bc44c487f
  Stored in directory: /Users/pushpakulkarni/Library/Caches/pip/wheels/8b/f1/7f/5c94f0a7a505ca1c81cd1d9208ae2064675d97582078e6c769
Successfully built wget
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 23.1/23.1 MB 18.0 MB/s eta 0:00:0000:0100:01
  Preparing metadata (setup.py) ... done
  Created wheel for gensim: filename=gensim-3.6.0-cp310-cp310-macosx_11_0_arm64.whl size=23220560 sha256=1c82bec6da49b4cb0d70c93710bf5e4bcfd35f3918504c246dfaad9c665630f4
  Stored in directory: /Users/pushpakulkarni/Library/Caches/pip/wheels/00/e8/47/96f55c3144a5ea3537f549f7a97607011f5004b9f13fa8dcc5
Successfully built gensim
  Attempting uninstall: gensim
    Found existing installation: gensim 4.3.1
    Uninstalling gensim-4.3.1:
      Successfully uninstalled gensim-4.3.1
ERROR: pip's dependency resolver does not currentl

            clang -c /private/var/folders/w7/x95pzq4s5s1f44mr9jmxg5_m0000gn/T/pip-install-smrwn1ta/blis_eadc796cd3494c2b9bab24f9001cd1ba/blis/_src/frame/1m/bli_l1m_oapi_ba.c -o /var/folders/w7/x95pzq4s5s1f44mr9jmxg5_m0000gn/T/tmpaz_wqgzf/bli_l1m_oapi_ba.o -O3 -fPIC -std=c99 -D_POSIX_C_SOURCE=200112L -DBLIS_VERSION_STRING="0.5.1" -DBLIS_IS_BUILDING_LIBRARY -Iinclude/darwin-x86_64 -I./frame/3/ -I./frame/ind/ukernels/ -I./frame/1m/ -I./frame/1f/ -I./frame/1/ -I./frame/include -I/private/var/folders/w7/x95pzq4s5s1f44mr9jmxg5_m0000gn/T/pip-install-smrwn1ta/blis_eadc796cd3494c2b9bab24f9001cd1ba/blis/_src/include/darwin-x86_64
            clang -c /private/var/folders/w7/x95pzq4s5s1f44mr9jmxg5_m0000gn/T/pip-install-smrwn1ta/blis_eadc796cd3494c2b9bab24f9001cd1ba/blis/_src/frame/1m/bli_l1m_oapi_ex.c -o /var/folders/w7/x95pzq4s5s1f44mr9jmxg5_m0000gn/T/tmpaz_wqgzf/bli_l1m_oapi_ex.o -O3 -fPIC -std=c99 -D_POSIX_C_SOURCE=200112L -DBLIS_VERSION_STRING="0.5.1" -DBLIS_IS_BUILDING_LIBRARY -Iinclude/darw

            clang -c /private/var/folders/w7/x95pzq4s5s1f44mr9jmxg5_m0000gn/T/pip-install-smrwn1ta/blis_eadc796cd3494c2b9bab24f9001cd1ba/blis/_src/ref_kernels/ind/bli_gemm3mh_ref.c -o /var/folders/w7/x95pzq4s5s1f44mr9jmxg5_m0000gn/T/tmp2x8ifw__/bli_gemm3mh_steamroller_ref.o -O3 -mfpmath=sse -mavx -mfma -march=bdver3 -mno-fma4 -mno-tbm -mno-xop -fPIC -std=c99 -D_POSIX_C_SOURCE=200112L -DBLIS_VERSION_STRING="0.5.1" -DBLIS_CNAME=steamroller -DBLIS_IS_BUILDING_LIBRARY -Iinclude/darwin-x86_64 -I./frame/3/ -I./frame/ind/ukernels/ -I./frame/1m/ -I./frame/1f/ -I./frame/1/ -I./frame/include -I/private/var/folders/w7/x95pzq4s5s1f44mr9jmxg5_m0000gn/T/pip-install-smrwn1ta/blis_eadc796cd3494c2b9bab24f9001cd1ba/blis/_src/include/darwin-x86_64
            clang -c /private/var/folders/w7/x95pzq4s5s1f44mr9jmxg5_m0000gn/T/pip-install-smrwn1ta/blis_eadc796cd3494c2b9bab24f9001cd1ba/blis/_src/ref_kernels/ind/bli_gemm4m1_ref.c -o /var/folders/w7/x95pzq4s5s1f44mr9jmxg5_m0000gn/T/tmp2x8ifw__/bli_gemm4m1_stea

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [3]:
import os
import wget
import gzip
import shutil

gn_vec_path = "GoogleNews-vectors-negative300.bin"
if not os.path.exists("GoogleNews-vectors-negative300.bin"):
    if not os.path.exists("/content/drive/MyDrive/GoogleNews-vectors-negative300.bin.gz"):
        #Downloading the reqired model
        if not os.path.exists("/content/drive/MyDrive/GoogleNews-vectors-negative300.bin.gz"):
            if not os.path.exists("GoogleNews-vectors-negative300.bin.gz"):
                wget.download("https://s3.amazonaws.com/dl4j-distribution/GoogleNews-vectors-negative300.bin.gz")
            gn_vec_zip_path = "GoogleNews-vectors-negative300.bin.gz"
        else:
            gn_vec_zip_path = "/content/drive/MyDrive/GoogleNews-vectors-negative300.bin.gz"
        #Extracting the required model
        with gzip.open(gn_vec_zip_path, 'rb') as f_in:
            with open(gn_vec_path, 'wb') as f_out:
                shutil.copyfileobj(f_in, f_out)
    else:
        gn_vec_path = "../Ch2/" + gn_vec_path

print(f"Model at {gn_vec_path}")

Model at ../Ch2/GoogleNews-vectors-negative300.bin


In [4]:
!pip install wget

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [5]:
!free -h

              total        used        free      shared  buff/cache   available
Mem:            12G        584M        9.6G        1.2M        2.6G         11G
Swap:            0B          0B          0B


In [6]:
!nvidia-smi -L

NVIDIA-SMI has failed because it couldn't communicate with the NVIDIA driver. Make sure that the latest NVIDIA driver is installed and running.



In [7]:
!lscpu |grep 'Model name'

Model name:          Intel(R) Xeon(R) CPU @ 2.20GHz


In [4]:
import warnings #This module ignores the various types of warnings generated
warnings.filterwarnings("ignore") 

import psutil #This module helps in retrieving information on running processes and system resource utilization
process = psutil.Process(os.getpid())
from psutil import virtual_memory
mem = virtual_memory()

import time #This module is used to calculate the time 

In [5]:
import os
import psutil #This module helps in retrieving information on running processes and system resource utilization
process = psutil.Process(os.getpid())
from psutil import virtual_memory
mem = virtual_memory()

import time #This module is used to calculate the time 
from gensim.models import Word2Vec, KeyedVectors
pretrainedpath = "/content/drive/MyDrive/GoogleNews-vectors-negative300.bin.gz"

#Load W2V model. This will take some time, but it is a one time effort! 
pre = process.memory_info().rss
print("Memory used in GB before Loading the Model: %0.2f"%float(pre/(10**9))) #Check memory usage before loading the model
print('-'*10)

start_time = time.time() #Start the timer
ttl = mem.total #Toal memory available

w2v_model = KeyedVectors.load_word2vec_format(pretrainedpath, binary=True) #load the model
print("%0.2f seconds taken to load"%float(time.time() - start_time)) #Calculate the total time elapsed since starting the timer
print('-'*10)

print('Finished loading Word2Vec')
print('-'*10)

post = process.memory_info().rss
print("Memory used in GB after Loading the Model: {:.2f}".format(float(post/(10**9)))) #Calculate the memory used after loading the model
print('-'*10)

print("Percentage increase in memory usage: {:.2f}% ".format(float((post/pre)*100))) #Percentage increase in memory after loading the model
print('-'*10)

print("Number of words in vocabulary: ",len(w2v_model.vocab)) #Number of words in the vocabulary. 

Memory used in GB before Loading the Model: 0.15
----------
140.71 seconds taken to load
----------
Finished loading Word2Vec
----------
Memory used in GB after Loading the Model: 5.09
----------
Percentage increase in memory usage: 3366.03% 
----------
Number of words in vocabulary:  3000000


In [ ]:
#Let us examine the model by knowing what the most similar words are, for a given word!
w2v_model.most_similar('beautiful')

[('gorgeous', 0.8353004455566406),
 ('lovely', 0.810693621635437),
 ('stunningly_beautiful', 0.7329413890838623),
 ('breathtakingly_beautiful', 0.7231341004371643),
 ('wonderful', 0.6854087114334106),
 ('fabulous', 0.6700063943862915),
 ('loveliest', 0.6612576246261597),
 ('prettiest', 0.6595001816749573),
 ('beatiful', 0.6593326330184937),
 ('magnificent', 0.6591402292251587)]

In [ ]:
#What is the vector representation for a word? 
w2v_model['computer']

array([ 1.07421875e-01, -2.01171875e-01,  1.23046875e-01,  2.11914062e-01,
       -9.13085938e-02,  2.16796875e-01, -1.31835938e-01,  8.30078125e-02,
        2.02148438e-01,  4.78515625e-02,  3.66210938e-02, -2.45361328e-02,
        2.39257812e-02, -1.60156250e-01, -2.61230469e-02,  9.71679688e-02,
       -6.34765625e-02,  1.84570312e-01,  1.70898438e-01, -1.63085938e-01,
       -1.09375000e-01,  1.49414062e-01, -4.65393066e-04,  9.61914062e-02,
        1.68945312e-01,  2.60925293e-03,  8.93554688e-02,  6.49414062e-02,
        3.56445312e-02, -6.93359375e-02, -1.46484375e-01, -1.21093750e-01,
       -2.27539062e-01,  2.45361328e-02, -1.24511719e-01, -3.18359375e-01,
       -2.20703125e-01,  1.30859375e-01,  3.66210938e-02, -3.63769531e-02,
       -1.13281250e-01,  1.95312500e-01,  9.76562500e-02,  1.26953125e-01,
        6.59179688e-02,  6.93359375e-02,  1.02539062e-02,  1.75781250e-01,
       -1.68945312e-01,  1.21307373e-03, -2.98828125e-01, -1.15234375e-01,
        5.66406250e-02, -

In [ ]:
#What if I am looking for a word that is not in this vocabulary?
w2v_model['practicalnlp']

KeyError: ignored

In [ ]:
!python -m spacy download en_core_web_md

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     |████████████████████████████████| 96.4 MB 1.2 MB/s 
  Created wheel for en-core-web-md: filename=en_core_web_md-2.2.5-py3-none-any.whl size=98051301 sha256=97f2bcbc937069ae4fe7256fe96e369569387c9e98d635b02245d969b30525ee
  Stored in directory: /tmp/pip-ephem-wheel-cache-lhd6kydn/wheels/69/c5/b8/4f1c029d89238734311b3269762ab2ee325a42da2ce8edb997
Successfully built en-core-web-md
✔ Download and installation successful
You can now load the model via spacy.load('en_core_web_md')


In [ ]:
import spacy

%time 
nlp = spacy.load('en_core_web_md')
# process a sentence using the model
mydoc = nlp("Canada is a large country")
#Get a vector for individual words
#print(doc[0].vector) #vector for 'Canada', the first word in the text 
print(mydoc.vector) #Averaged vector for the entire sentence

CPU times: user 2 µs, sys: 0 ns, total: 2 µs
Wall time: 7.63 µs
[-1.12055197e-01  2.26087615e-01 -5.15111461e-02 -1.21812008e-01
  4.13958639e-01 -8.56475979e-02 -2.84600933e-03 -2.26096585e-01
  6.98113963e-02  2.27946019e+00 -4.49774921e-01 -6.39050007e-02
 -1.80326015e-01 -8.79765972e-02  9.93399299e-04 -1.57384202e-01
 -1.23817801e-01  1.54990411e+00  2.00794004e-02  1.38399601e-01
 -1.48897991e-01 -2.23025799e-01 -1.48171991e-01  4.68924567e-02
 -3.17026004e-02  1.19096041e-02 -6.10985979e-02  9.57068056e-02
  9.37099904e-02  1.70955807e-01 -9.29740071e-03  7.88536817e-02
  1.74508005e-01 -1.04450598e-01  1.04872189e-01 -1.16961405e-01
  6.23028055e-02 -2.23016590e-01 -1.44107476e-01 -2.03423887e-01
  2.61404991e-01  2.43404001e-01  1.51980996e-01 -1.12484001e-01
  1.18055798e-01 -9.51323956e-02  8.66319984e-02 -2.54322797e-01
  3.84932049e-02  1.18278004e-01 -3.21602583e-01  3.73764008e-01
  1.13018408e-01 -8.05834010e-02  1.84921592e-01  9.38879885e-03
  1.22166201e-01 -3.242880

#Exercises

Programming Assignment 7

1. Use a pre-trained word embedding model and find out the embedding of some random words and their most similar words . 
2. Create a toy corpus  having four sentences with maximum 5 words into each sentence. Peform Text cleaning and preprocessing and use gensim to create a CBoW and Skipgram model and then trained them on your toy corpus. Observe the embeddings 
3. Create a corpus  by taking a raw text. Peform required text cleaning and preprocessing and use gensim to create a CBoW and Skipgram model and then trained them your corpus. Observe the embeddings 

In [ ]:
#TASK-1

In [ ]:
w2v_model.most_similar('smart')

[('intelligent', 0.6495277881622314),
 ('dumb', 0.5792694687843323),
 ('smartest', 0.5717369318008423),
 ('savvy', 0.5674132108688354),
 ('clever', 0.5656732320785522),
 ('smarter', 0.5632593631744385),
 ('shrewd', 0.5591170191764832),
 ('IPICO_produces', 0.5158056020736694),
 ('economic_certainty_Kracmer', 0.5094711184501648),
 ('Smart', 0.4991162121295929)]

In [ ]:
w2v_model['smart']

array([ 1.04492188e-01,  1.37939453e-02, -6.03027344e-02,  9.96093750e-02,
       -1.47460938e-01, -5.61523438e-02, -4.93164062e-02,  9.58251953e-03,
       -6.53076172e-03, -9.52148438e-02, -3.44848633e-03, -1.26953125e-01,
       -9.47265625e-02, -2.12890625e-01,  6.83593750e-02, -1.34765625e-01,
        4.56542969e-02,  4.93164062e-02, -2.16796875e-01,  2.01416016e-03,
        1.97753906e-02, -1.79443359e-02,  1.30859375e-01,  1.12792969e-01,
        1.11816406e-01,  2.51953125e-01, -1.21093750e-01,  1.40625000e-01,
       -9.86328125e-02, -2.33398438e-01, -8.39843750e-02,  1.66015625e-01,
        1.55273438e-01,  1.12304688e-01,  2.87109375e-01,  6.64062500e-02,
       -6.07910156e-02,  1.69677734e-02, -4.15039062e-02,  7.42187500e-02,
        1.62109375e-01, -1.81640625e-01,  3.08593750e-01, -5.83496094e-02,
       -3.10058594e-02,  9.86328125e-02, -1.17675781e-01, -2.71484375e-01,
        7.22656250e-02, -5.37109375e-02, -4.73632812e-02,  1.54296875e-01,
       -5.81054688e-02, -

In [ ]:
w2v_model.most_similar('hot')

[('Hot', 0.6659734845161438),
 ('hottest', 0.6050904989242554),
 ('hotter', 0.5794501900672913),
 ('CHEFS_Chefs', 0.5725089907646179),
 ('sizzling', 0.5456805229187012),
 ('scorching', 0.5294975638389587),
 ('teardrop_shaped_VTA', 0.5230512619018555),
 ('cool', 0.5151149034500122),
 ('briquette_iron', 0.5066832304000854),
 ('heated', 0.5010007619857788)]

In [ ]:
w2v_model['hot']

array([-0.00750732,  0.00817871,  0.07373047,  0.0859375 , -0.04418945,
       -0.13769531,  0.09619141, -0.20019531, -0.20214844,  0.19433594,
        0.03710938, -0.15429688,  0.13085938, -0.08203125, -0.10058594,
        0.15625   ,  0.05908203, -0.18261719,  0.06347656, -0.20410156,
        0.01843262,  0.13671875, -0.10205078, -0.09667969, -0.05493164,
        0.10107422,  0.03198242,  0.09960938, -0.1484375 ,  0.04174805,
        0.08007812,  0.171875  ,  0.22460938, -0.12695312, -0.22460938,
       -0.12207031, -0.11816406,  0.02868652,  0.14453125, -0.07324219,
        0.06079102, -0.41210938,  0.02819824, -0.1015625 ,  0.015625  ,
       -0.20605469,  0.109375  , -0.07421875,  0.20703125,  0.08544922,
        0.13574219,  0.421875  , -0.07763672,  0.1953125 , -0.25390625,
        0.0546875 ,  0.30664062, -0.03173828,  0.09814453,  0.03015137,
       -0.09716797,  0.00543213, -0.26367188, -0.09375   ,  0.14355469,
        0.04150391, -0.3125    , -0.50390625,  0.00127411,  0.04

In [ ]:
w2v_model.most_similar('sweet')

[('sweetness', 0.6216812133789062),
 ('sweetest', 0.6207044720649719),
 ('caramelly', 0.6009937524795532),
 ('syrupy_sweet', 0.5979651808738708),
 ('yummy', 0.5963824391365051),
 ('buttery', 0.593994140625),
 ('tooth_achingly', 0.5892406105995178),
 ('fan_Rosangela_Pereira', 0.5891238451004028),
 ('delicious', 0.5866374969482422),
 ('fruity', 0.5804054737091064)]

In [ ]:
w2v_model['sweet']

array([ 0.08691406,  0.0300293 , -0.04589844,  0.11767578,  0.02490234,
       -0.07421875,  0.31640625, -0.00331116, -0.07226562,  0.23925781,
       -0.2734375 , -0.1640625 , -0.24121094, -0.0201416 ,  0.06176758,
        0.2890625 , -0.06298828, -0.01599121, -0.01635742, -0.06787109,
       -0.06591797,  0.25976562,  0.10351562, -0.05224609,  0.11035156,
       -0.13671875, -0.11621094, -0.06640625, -0.03442383,  0.10058594,
        0.03271484,  0.12988281,  0.140625  ,  0.17773438, -0.19335938,
       -0.13378906,  0.05175781, -0.13574219,  0.02490234,  0.07910156,
        0.19628906, -0.3203125 ,  0.09423828, -0.02258301,  0.07519531,
       -0.24804688,  0.09179688,  0.24707031,  0.11621094,  0.10449219,
       -0.17871094,  0.2734375 ,  0.19335938,  0.18652344,  0.01953125,
        0.15917969, -0.04516602, -0.06494141, -0.109375  , -0.21582031,
       -0.31835938,  0.12597656, -0.21972656,  0.15039062, -0.00408936,
       -0.06591797, -0.02087402, -0.3515625 ,  0.07666016,  0.26

In [ ]:
w2v_model.most_similar('friend')

[('pal', 0.7476358413696289),
 ('friends', 0.7098034620285034),
 ('buddy', 0.6972494125366211),
 ('dear_friend', 0.6960037350654602),
 ('acquaintance', 0.6843010187149048),
 ('cousin', 0.6713200807571411),
 ('girlfriend', 0.6226294040679932),
 ('colleague', 0.6204894185066223),
 ('uncle', 0.6120923161506653),
 ('roommate', 0.611858069896698)]

In [ ]:
w2v_model['friend']

array([ 0.07080078, -0.21386719,  0.15332031,  0.09423828, -0.03442383,
        0.43359375, -0.16503906, -0.05786133,  0.17578125, -0.08203125,
        0.24511719, -0.19335938, -0.0255127 , -0.09619141, -0.125     ,
        0.02575684,  0.16796875, -0.03759766,  0.09472656, -0.04760742,
        0.20605469,  0.31835938,  0.15917969, -0.17089844,  0.09033203,
       -0.1640625 , -0.15234375,  0.3125    ,  0.06298828, -0.24902344,
        0.15625   , -0.04516602, -0.12890625, -0.00686646, -0.02160645,
        0.14453125,  0.2734375 ,  0.12695312,  0.10742188,  0.11376953,
        0.14355469, -0.00173187,  0.22851562, -0.03515625,  0.17089844,
        0.04516602, -0.07958984, -0.08886719, -0.01342773, -0.09667969,
       -0.12597656,  0.10595703,  0.15332031, -0.03808594,  0.02246094,
        0.01428223, -0.03295898,  0.20703125, -0.03417969,  0.02233887,
        0.00244141,  0.13476562, -0.01403809,  0.13378906,  0.0201416 ,
        0.14746094,  0.00759888, -0.18652344,  0.16113281,  0.10

In [ ]:
#TASK-2

In [6]:
sent1 = 'Hello this is Millie'
sent2 = 'Millie and Shreyas are friends'
sent3 = 'Millie likes Physics'
sent4 = 'Shreyas likes NLP'

data = [sent1.split(), sent2.split(), sent3.split(), sent4.split()]
values = data[0]+data[1]+data[2]+data[3]
our_data = values
print(values)
print("The data: ",values)

['Hello', 'this', 'is', 'Millie', 'Millie', 'and', 'Shreyas', 'are', 'friends', 'Millie', 'likes', 'Physics', 'Shreyas', 'likes', 'NLP']
The data:  ['Hello', 'this', 'is', 'Millie', 'Millie', 'and', 'Shreyas', 'are', 'friends', 'Millie', 'likes', 'Physics', 'Shreyas', 'likes', 'NLP']


In [7]:
import re
import pandas as pd
import nltk
import string
from nltk.corpus import stopwords
nltk.download('stopwords')
nltk.download('punkt')


def clean(doc): # doc is a string of text
    #doc = [x.replace("</br>", " ") for x in doc] # This text contains a lot of <br/> tags.      [x.replace("</br>", " ") for x in doc] doc.replace("</br>", " ")
    #doc = "".join([char for char in doc if char not in string.punctuation and not char.isdigit()])
    #doc = " ".join([token for token in doc.split() if token not in stopwords])
    # remove punctuation and numbers
    #initialize stopwords
    stop_words_nltk = set(stopwords.words('english'))
    #print(stop_words_nltk)

    #stopword removal
    corpus2 = [i for i in doc if not i in stop_words_nltk]
    print("Tokenized corpus without stopwords:",corpus2)
    return corpus2
our_data = clean(our_data)   


/usr/local/lib/python3.7/dist-packages/sklearn/feature_extraction/image.py:167: DeprecationWarning: `np.int` is a deprecated alias for the builtin `int`. To silence this warning, use `int` by itself. Doing this will not modify any behavior and is safe. When replacing `np.int`, you may wish to use e.g. `np.int64` or `np.int32` to specify the precision. If you wish to review your current use, check the release note link for additional information.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  dtype=np.int):
/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/least_angle.py:30: DeprecationWarning: `np.float` is a deprecated alias for the builtin `float`. To silence this warning, use `float` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.float64` here.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdo

Tokenized corpus without stopwords: ['Hello', 'Millie', 'Millie', 'Shreyas', 'friends', 'Millie', 'likes', 'Physics', 'Shreyas', 'likes', 'NLP']


In [8]:
def extractL(lst):
    return list(map(lambda el:[el], lst))

our_data = extractL(our_data)    
our_data

[['Hello'],
 ['Millie'],
 ['Millie'],
 ['Shreyas'],
 ['friends'],
 ['Millie'],
 ['likes'],
 ['Physics'],
 ['Shreyas'],
 ['likes'],
 ['NLP']]

In [9]:
model = Word2Vec(our_data, min_count=1,size= 50,workers=3, window =3, sg = 0) #sg=0 default CBOW for skipgram set to 1

model2 = Word2Vec(our_data, min_count=1,size= 50,workers=5, window =2, sg = 1)

In [ ]:
model['NLP']

array([ 1.3146516e-03,  8.7559475e-03, -2.5728724e-03,  1.1237229e-04,
        1.8875578e-03, -5.0069867e-03,  6.7687416e-03, -9.9992510e-03,
       -4.1280923e-04, -2.2862162e-03, -9.7010909e-03,  6.9062519e-03,
        9.5691029e-03, -3.0840512e-03,  4.4608414e-03, -1.7764017e-03,
       -8.2435962e-03, -3.2646020e-03, -9.9094855e-03, -9.4692912e-03,
        8.7401075e-03,  2.2220907e-03, -8.7813307e-03, -4.3473379e-03,
       -3.9097221e-04,  4.8672589e-03,  1.5777374e-03, -2.6316158e-04,
       -8.8995704e-03,  9.8497635e-03,  4.2200121e-03,  3.4971384e-03,
        2.8517384e-03,  8.2129147e-03,  8.1832148e-03,  1.3322410e-03,
        1.2316037e-05,  5.6195138e-03,  2.0136524e-03, -4.8742563e-04,
       -7.2859200e-03, -9.2956088e-03, -6.6227694e-03,  1.2034532e-03,
        5.1734657e-03,  5.3558438e-03,  6.4542238e-03, -5.8507882e-03,
        2.3829192e-03,  1.3914639e-03], dtype=float32)

In [ ]:
model['Hello']

array([-7.7724608e-04, -8.9907041e-03,  7.3977425e-03, -7.7399369e-03,
        1.3372888e-03,  8.3685564e-03, -5.5546416e-03, -6.6069402e-03,
       -5.7461751e-03,  9.4375963e-04, -9.0449052e-03,  3.1865919e-03,
        3.0637851e-03, -6.6448567e-03,  8.2862442e-03,  6.0003200e-03,
        5.0570504e-03, -8.4179537e-03, -2.0104467e-03, -5.4620481e-03,
       -4.0113931e-03, -2.8337825e-05, -7.0015886e-03, -4.2749051e-04,
       -1.3846685e-03, -5.6160851e-03,  2.1336376e-04,  1.8454179e-03,
       -2.9860779e-03,  6.1147953e-03,  1.4553561e-03, -3.2206960e-03,
        5.7934150e-03, -2.4389478e-03,  2.5247803e-03, -5.2044098e-03,
       -4.4597303e-03,  4.4761668e-03, -8.7460503e-05, -5.0890315e-03,
        3.7272116e-03,  8.4023876e-03, -8.7595712e-03,  2.7472220e-04,
       -4.1439533e-03, -5.2641677e-03,  5.2119493e-03, -9.1280788e-03,
       -1.8064809e-03,  2.2712999e-03], dtype=float32)

In [ ]:
model['Millie']

array([ 1.8134555e-05, -3.9992682e-04,  6.0865846e-03,  4.1675293e-03,
        3.6436340e-03, -8.8238623e-03, -8.3572650e-03,  5.8367532e-03,
        5.8944155e-03, -4.9712374e-03, -3.1488489e-03, -1.8000426e-03,
       -9.9649290e-03, -5.7536447e-03, -2.1811598e-03,  8.4818210e-03,
       -6.0961382e-03,  1.5411370e-03, -4.9108048e-03,  7.9836808e-03,
        2.2749444e-03,  6.8851812e-03,  6.2341075e-03, -4.4021336e-03,
        5.3096451e-03,  1.8955993e-03,  2.5525915e-03, -9.6495869e-03,
        3.0002154e-03,  9.3029039e-03,  7.9389010e-03, -8.5610524e-03,
       -4.6501551e-03, -7.5360546e-03,  5.1105209e-04, -7.3738475e-03,
       -7.4059195e-03, -2.8706687e-03, -7.4814255e-03,  3.2218655e-03,
        8.9386152e-03, -5.9583876e-03,  7.4548036e-04, -4.1315001e-03,
       -9.9905375e-03,  1.0565391e-03, -2.9333830e-03, -7.6596471e-03,
        3.9764252e-03,  2.0230883e-03], dtype=float32)

In [ ]:
model.most_similar('Millie')

[('likes', 0.04230295866727829),
 ('Hello', 0.0013784740585833788),
 ('Physics', -0.0656752660870552),
 ('friends', -0.08807007968425751),
 ('NLP', -0.11610457301139832),
 ('Shreyas', -0.17405158281326294)]

In [ ]:
model['Shreyas']

array([-3.1007377e-03,  8.8250246e-03, -1.7164573e-03, -2.8489414e-04,
       -6.6798795e-03,  4.2225556e-03,  1.2480994e-03, -8.7552397e-03,
       -1.9373518e-03,  2.5600396e-04,  8.9714229e-03,  5.1522618e-03,
        9.8919717e-04,  6.7502185e-04, -9.8543344e-03,  3.4552913e-03,
       -2.5372670e-03, -6.0131126e-03,  4.7803996e-03, -5.5559492e-03,
        5.3254263e-03,  7.1089347e-03,  2.4110294e-04,  6.4840536e-03,
       -4.1129002e-03,  6.4601186e-03, -5.4510897e-03,  5.4066246e-03,
       -9.9154869e-03,  6.6665183e-03, -4.2058756e-03, -9.9315345e-03,
        1.2303403e-03,  2.4500587e-03, -4.2300988e-03, -7.2377077e-03,
        5.5619841e-03,  8.2554650e-03,  9.0061547e-03,  5.5886339e-03,
       -1.4477929e-03,  5.4206932e-03, -2.5495712e-05, -1.1280694e-06,
        8.3391694e-03, -1.1670056e-03, -4.5022862e-03, -4.2864438e-03,
        8.2748979e-03,  8.9964010e-03], dtype=float32)

In [ ]:
model.most_similar('Shreyas')

[('friends', 0.2498396635055542),
 ('NLP', 0.18875345587730408),
 ('Physics', 0.05735952407121658),
 ('Hello', 0.016872897744178772),
 ('likes', 0.010980001650750637),
 ('Millie', -0.17405156791210175)]

In [ ]:
model['NLP']

array([ 1.3146516e-03,  8.7559475e-03, -2.5728724e-03,  1.1237229e-04,
        1.8875578e-03, -5.0069867e-03,  6.7687416e-03, -9.9992510e-03,
       -4.1280923e-04, -2.2862162e-03, -9.7010909e-03,  6.9062519e-03,
        9.5691029e-03, -3.0840512e-03,  4.4608414e-03, -1.7764017e-03,
       -8.2435962e-03, -3.2646020e-03, -9.9094855e-03, -9.4692912e-03,
        8.7401075e-03,  2.2220907e-03, -8.7813307e-03, -4.3473379e-03,
       -3.9097221e-04,  4.8672589e-03,  1.5777374e-03, -2.6316158e-04,
       -8.8995704e-03,  9.8497635e-03,  4.2200121e-03,  3.4971384e-03,
        2.8517384e-03,  8.2129147e-03,  8.1832148e-03,  1.3322410e-03,
        1.2316037e-05,  5.6195138e-03,  2.0136524e-03, -4.8742563e-04,
       -7.2859200e-03, -9.2956088e-03, -6.6227694e-03,  1.2034532e-03,
        5.1734657e-03,  5.3558438e-03,  6.4542238e-03, -5.8507882e-03,
        2.3829192e-03,  1.3914639e-03], dtype=float32)

In [ ]:
model.most_similar('NLP')

[('Shreyas', 0.18875345587730408),
 ('Hello', 0.1579628437757492),
 ('likes', 0.08836456388235092),
 ('friends', 0.06421137601137161),
 ('Millie', -0.11610454320907593),
 ('Physics', -0.33987927436828613)]

In [18]:
temp2=our_data
res = [''.join(ele) for ele in temp2]
str1 =''
for ele in res:
        str1 += ele
str1        

'HelloMillieMillieShreyasfriendsMillielikesPhysicsShreyaslikesNLP'

In [15]:
import numpy as np
def get_mean_vector(word2vec_model, words):
    # remove out-of-vocabulary words
    words = [word for word in words if word in word2vec_model]
    if len(words) >= 1:
        return np.mean(word2vec_model[words], axis=0)
    else:
        return []

In [56]:
word_embedding_avg_collectivei = []   #Averaged vector for the entire sentence for CBOW
for doc in our_data:
    vec = get_mean_vector(model, doc)
    word_embedding_avg_collectivei.append(vec)
    if len(vec) > 0:
      print('Dimension of vector for each word:',vec.shape)

word_embedding_avgi = np.mean(word_embedding_avg_collectivei, axis=0)     
print() 
print('Dimension of vector for overall corpus:',word_embedding_avgi.shape)

print('Vector for overall corpus:',word_embedding_avgi)

Dimension of vector for each word: (50,)
Dimension of vector for each word: (50,)
Dimension of vector for each word: (50,)
Dimension of vector for each word: (50,)
Dimension of vector for each word: (50,)
Dimension of vector for each word: (50,)
Dimension of vector for each word: (50,)
Dimension of vector for each word: (50,)
Dimension of vector for each word: (50,)
Dimension of vector for each word: (50,)
Dimension of vector for each word: (50,)

Dimension of vector for overall corpus: (50,)
Vector for overall corpus: [-0.00042292 -0.00128124  0.0016713  -0.00100073  0.00264984 -0.00056274
 -0.0018508   0.00528235  0.00074241 -0.00030788  0.00358317  0.00211925
  0.00028895  0.00125559  0.00045029 -0.00221913 -0.00109371 -0.00282404
 -0.00254699 -0.00171106 -0.00148378 -0.00040002 -0.00417252  0.0016632
 -0.00242404 -0.00181534  0.00011485  0.00287269  0.00050896  0.00212136
  0.00023629  0.00192432  0.00266776 -0.00164606  0.00342413  0.00275134
  0.00014774 -0.00225084  0.0016815   

In [ ]:
#for skipgrams

In [ ]:
model2['NLP']

array([ 1.3146516e-03,  8.7559475e-03, -2.5728724e-03,  1.1237229e-04,
        1.8875578e-03, -5.0069867e-03,  6.7687416e-03, -9.9992510e-03,
       -4.1280923e-04, -2.2862162e-03, -9.7010909e-03,  6.9062519e-03,
        9.5691029e-03, -3.0840512e-03,  4.4608414e-03, -1.7764017e-03,
       -8.2435962e-03, -3.2646020e-03, -9.9094855e-03, -9.4692912e-03,
        8.7401075e-03,  2.2220907e-03, -8.7813307e-03, -4.3473379e-03,
       -3.9097221e-04,  4.8672589e-03,  1.5777374e-03, -2.6316158e-04,
       -8.8995704e-03,  9.8497635e-03,  4.2200121e-03,  3.4971384e-03,
        2.8517384e-03,  8.2129147e-03,  8.1832148e-03,  1.3322410e-03,
        1.2316037e-05,  5.6195138e-03,  2.0136524e-03, -4.8742563e-04,
       -7.2859200e-03, -9.2956088e-03, -6.6227694e-03,  1.2034532e-03,
        5.1734657e-03,  5.3558438e-03,  6.4542238e-03, -5.8507882e-03,
        2.3829192e-03,  1.3914639e-03], dtype=float32)

In [ ]:
model2.most_similar('NLP')

[('Shreyas', 0.18875345587730408),
 ('Hello', 0.1579628437757492),
 ('likes', 0.08836456388235092),
 ('friends', 0.06421137601137161),
 ('Millie', -0.11610454320907593),
 ('Physics', -0.33987927436828613)]

In [ ]:
model2['Shreyas']

array([-3.1007377e-03,  8.8250246e-03, -1.7164573e-03, -2.8489414e-04,
       -6.6798795e-03,  4.2225556e-03,  1.2480994e-03, -8.7552397e-03,
       -1.9373518e-03,  2.5600396e-04,  8.9714229e-03,  5.1522618e-03,
        9.8919717e-04,  6.7502185e-04, -9.8543344e-03,  3.4552913e-03,
       -2.5372670e-03, -6.0131126e-03,  4.7803996e-03, -5.5559492e-03,
        5.3254263e-03,  7.1089347e-03,  2.4110294e-04,  6.4840536e-03,
       -4.1129002e-03,  6.4601186e-03, -5.4510897e-03,  5.4066246e-03,
       -9.9154869e-03,  6.6665183e-03, -4.2058756e-03, -9.9315345e-03,
        1.2303403e-03,  2.4500587e-03, -4.2300988e-03, -7.2377077e-03,
        5.5619841e-03,  8.2554650e-03,  9.0061547e-03,  5.5886339e-03,
       -1.4477929e-03,  5.4206932e-03, -2.5495712e-05, -1.1280694e-06,
        8.3391694e-03, -1.1670056e-03, -4.5022862e-03, -4.2864438e-03,
        8.2748979e-03,  8.9964010e-03], dtype=float32)

In [ ]:
model2.most_similar('Shreyas')

[('friends', 0.2498396635055542),
 ('NLP', 0.18875345587730408),
 ('Physics', 0.05735952407121658),
 ('Hello', 0.016872897744178772),
 ('likes', 0.010980001650750637),
 ('Millie', -0.17405156791210175)]

In [ ]:
model2['Millie']

array([ 1.8134555e-05, -3.9992682e-04,  6.0865846e-03,  4.1675293e-03,
        3.6436340e-03, -8.8238623e-03, -8.3572650e-03,  5.8367532e-03,
        5.8944155e-03, -4.9712374e-03, -3.1488489e-03, -1.8000426e-03,
       -9.9649290e-03, -5.7536447e-03, -2.1811598e-03,  8.4818210e-03,
       -6.0961382e-03,  1.5411370e-03, -4.9108048e-03,  7.9836808e-03,
        2.2749444e-03,  6.8851812e-03,  6.2341075e-03, -4.4021336e-03,
        5.3096451e-03,  1.8955993e-03,  2.5525915e-03, -9.6495869e-03,
        3.0002154e-03,  9.3029039e-03,  7.9389010e-03, -8.5610524e-03,
       -4.6501551e-03, -7.5360546e-03,  5.1105209e-04, -7.3738475e-03,
       -7.4059195e-03, -2.8706687e-03, -7.4814255e-03,  3.2218655e-03,
        8.9386152e-03, -5.9583876e-03,  7.4548036e-04, -4.1315001e-03,
       -9.9905375e-03,  1.0565391e-03, -2.9333830e-03, -7.6596471e-03,
        3.9764252e-03,  2.0230883e-03], dtype=float32)

In [57]:
word_embedding_avg_collectiveii = []            #Averaged vector for the entire sentence-skipgram
for doc in our_data:
    vec = get_mean_vector(model2, doc)
    word_embedding_avg_collectiveii.append(vec)
    if len(vec) > 0:
      print('Dimension of vector for each word:',vec.shape)

word_embedding_avgii = np.mean(word_embedding_avg_collectiveii, axis=0)     
print() 
print('Dimension of vector for overall corpus:',word_embedding_avgii.shape)

print('Vector for overall corpus:',word_embedding_avgii) 

Dimension of vector for each word: (50,)
Dimension of vector for each word: (50,)
Dimension of vector for each word: (50,)
Dimension of vector for each word: (50,)
Dimension of vector for each word: (50,)
Dimension of vector for each word: (50,)
Dimension of vector for each word: (50,)
Dimension of vector for each word: (50,)
Dimension of vector for each word: (50,)
Dimension of vector for each word: (50,)
Dimension of vector for each word: (50,)

Dimension of vector for overall corpus: (50,)
Vector for overall corpus: [-0.00042292 -0.00128124  0.0016713  -0.00100073  0.00264984 -0.00056274
 -0.0018508   0.00528235  0.00074241 -0.00030788  0.00358317  0.00211925
  0.00028895  0.00125559  0.00045029 -0.00221913 -0.00109371 -0.00282404
 -0.00254699 -0.00171106 -0.00148378 -0.00040002 -0.00417252  0.0016632
 -0.00242404 -0.00181534  0.00011485  0.00287269  0.00050896  0.00212136
  0.00023629  0.00192432  0.00266776 -0.00164606  0.00342413  0.00275134
  0.00014774 -0.00225084  0.0016815   

In [46]:
#Doc2Vec
temp = our_data
temp2 = []
for i in temp:
    for j in i:
        temp2.append(j)
import gensim
docmodel = gensim.models.doc2vec.Doc2Vec(vector_size=50, min_count=2, epochs=40)

#building vocab
#docmodel.build_vocab([s.encode('utf-8').split() for s in temp2])


sentences = [gensim.models.doc2vec.TaggedDocument(sentence, 'tag') for sentence in temp2]
docmodel.build_vocab(sentences)

In [44]:
#temp3 = []
#temp3.append(temp2)
print(temp3)
print(temp2)

[['Hello', 'Millie', 'Millie', 'Shreyas', 'friends', 'Millie', 'likes', 'Physics', 'Shreyas', 'likes', 'NLP']]
['Hello', 'Millie', 'Millie', 'Shreyas', 'friends', 'Millie', 'likes', 'Physics', 'Shreyas', 'likes', 'NLP']


In [48]:
docmodel.train(sentences, total_examples=docmodel.corpus_count, epochs=docmodel.epochs)

In [58]:
vector = docmodel.infer_vector(temp2)
print(vector)                             #dim 50

[-0.00893469  0.00201269 -0.00633589 -0.00504674 -0.00998972 -0.00796403
 -0.00760187 -0.00552995  0.00727674 -0.00148645 -0.00219974 -0.00134809
  0.00432421 -0.00876796  0.00342459  0.00795252 -0.00072752  0.00848563
  0.00196946  0.00776178 -0.00220877  0.00096645  0.00977787 -0.00434421
 -0.00197775 -0.00749059  0.00766959  0.00336002  0.00862157  0.00630099
  0.00557941 -0.00040277  0.00151306  0.00681429  0.00199014  0.00178147
  0.00159094  0.00194351 -0.00194607  0.00822841 -0.0087912  -0.00631416
  0.00248155  0.0084832  -0.00275759  0.00308281 -0.00087546  0.00982781
  0.00596808 -0.00100695]


In [61]:
from scipy import spatial

dataSetI = word_embedding_avgi
dataSetII = word_embedding_avgii                                                    #−1  value will indicate strongly opposite vectors
dataSetIII = vector                                                                 #0 independent (orthogonal) vectors
result1 = 1 - spatial.distance.cosine(word_embedding_avgi, word_embedding_avgii)    #1 similar (positive co-linear) vectors. Intermediate values are used to assess 
result1                                                                               #the degree of similarity.

1

In [62]:
result2 = 1 - spatial.distance.cosine(word_embedding_avgi, vector)
result2

-0.22039258480072021

In [25]:
#TASK-3
corpusn = "NLP drives computer programs that translate text from one language to another, respond to spoken commands, and summarize large volumes of text rapidly—even in real time. There’s a good chance you’ve interacted with NLP in the form of voice-operated GPS systems, digital assistants, speech-to-text dictation software, customer service chatbots, and other consumer conveniences. But NLP also plays a growing role in enterprise solutions that help streamline business operations, increase employee productivity, and simplify mission-critical business processes."
values = corpusn.split()
stop_words_nltk = set(stopwords.words('english'))
values = [ i for i in values if not i in stop_words_nltk]

values

['NLP',
 'drives',
 'computer',
 'programs',
 'translate',
 'text',
 'one',
 'language',
 'another,',
 'respond',
 'spoken',
 'commands,',
 'summarize',
 'large',
 'volumes',
 'text',
 'rapidly—even',
 'real',
 'time.',
 'There’s',
 'good',
 'chance',
 'you’ve',
 'interacted',
 'NLP',
 'form',
 'voice-operated',
 'GPS',
 'systems,',
 'digital',
 'assistants,',
 'speech-to-text',
 'dictation',
 'software,',
 'customer',
 'service',
 'chatbots,',
 'consumer',
 'conveniences.',
 'But',
 'NLP',
 'also',
 'plays',
 'growing',
 'role',
 'enterprise',
 'solutions',
 'help',
 'streamline',
 'business',
 'operations,',
 'increase',
 'employee',
 'productivity,',
 'simplify',
 'mission-critical',
 'business',
 'processes.']

In [26]:
values = extractL(values)

In [27]:
values

[['NLP'],
 ['drives'],
 ['computer'],
 ['programs'],
 ['translate'],
 ['text'],
 ['one'],
 ['language'],
 ['another,'],
 ['respond'],
 ['spoken'],
 ['commands,'],
 ['summarize'],
 ['large'],
 ['volumes'],
 ['text'],
 ['rapidly—even'],
 ['real'],
 ['time.'],
 ['There’s'],
 ['good'],
 ['chance'],
 ['you’ve'],
 ['interacted'],
 ['NLP'],
 ['form'],
 ['voice-operated'],
 ['GPS'],
 ['systems,'],
 ['digital'],
 ['assistants,'],
 ['speech-to-text'],
 ['dictation'],
 ['software,'],
 ['customer'],
 ['service'],
 ['chatbots,'],
 ['consumer'],
 ['conveniences.'],
 ['But'],
 ['NLP'],
 ['also'],
 ['plays'],
 ['growing'],
 ['role'],
 ['enterprise'],
 ['solutions'],
 ['help'],
 ['streamline'],
 ['business'],
 ['operations,'],
 ['increase'],
 ['employee'],
 ['productivity,'],
 ['simplify'],
 ['mission-critical'],
 ['business'],
 ['processes.']]

In [28]:
model3 = Word2Vec(values, min_count=1,size= 50,workers=3, window =3, sg = 0) #sg=0 default CBOW for skipgram set to 1

model4 = Word2Vec(values, min_count=1,size= 50,workers=7, window =2, sg = 1)

In [ ]:
model3['NLP']

array([ 1.3146516e-03,  8.7559475e-03, -2.5728724e-03,  1.1237229e-04,
        1.8875578e-03, -5.0069867e-03,  6.7687416e-03, -9.9992510e-03,
       -4.1280923e-04, -2.2862162e-03, -9.7010909e-03,  6.9062519e-03,
        9.5691029e-03, -3.0840512e-03,  4.4608414e-03, -1.7764017e-03,
       -8.2435962e-03, -3.2646020e-03, -9.9094855e-03, -9.4692912e-03,
        8.7401075e-03,  2.2220907e-03, -8.7813307e-03, -4.3473379e-03,
       -3.9097221e-04,  4.8672589e-03,  1.5777374e-03, -2.6316158e-04,
       -8.8995704e-03,  9.8497635e-03,  4.2200121e-03,  3.4971384e-03,
        2.8517384e-03,  8.2129147e-03,  8.1832148e-03,  1.3322410e-03,
        1.2316037e-05,  5.6195138e-03,  2.0136524e-03, -4.8742563e-04,
       -7.2859200e-03, -9.2956088e-03, -6.6227694e-03,  1.2034532e-03,
        5.1734657e-03,  5.3558438e-03,  6.4542238e-03, -5.8507882e-03,
        2.3829192e-03,  1.3914639e-03], dtype=float32)

In [ ]:
model4['NLP']

array([ 1.3146516e-03,  8.7559475e-03, -2.5728724e-03,  1.1237229e-04,
        1.8875578e-03, -5.0069867e-03,  6.7687416e-03, -9.9992510e-03,
       -4.1280923e-04, -2.2862162e-03, -9.7010909e-03,  6.9062519e-03,
        9.5691029e-03, -3.0840512e-03,  4.4608414e-03, -1.7764017e-03,
       -8.2435962e-03, -3.2646020e-03, -9.9094855e-03, -9.4692912e-03,
        8.7401075e-03,  2.2220907e-03, -8.7813307e-03, -4.3473379e-03,
       -3.9097221e-04,  4.8672589e-03,  1.5777374e-03, -2.6316158e-04,
       -8.8995704e-03,  9.8497635e-03,  4.2200121e-03,  3.4971384e-03,
        2.8517384e-03,  8.2129147e-03,  8.1832148e-03,  1.3322410e-03,
        1.2316037e-05,  5.6195138e-03,  2.0136524e-03, -4.8742563e-04,
       -7.2859200e-03, -9.2956088e-03, -6.6227694e-03,  1.2034532e-03,
        5.1734657e-03,  5.3558438e-03,  6.4542238e-03, -5.8507882e-03,
        2.3829192e-03,  1.3914639e-03], dtype=float32)

In [ ]:
model3['business']

array([ 0.00311132,  0.00734582, -0.0038429 ,  0.00668994, -0.00175607,
       -0.00351088,  0.00892915, -0.00773621, -0.00191435,  0.00024624,
       -0.00823071, -0.00186822,  0.00824068, -0.0065991 , -0.00107715,
       -0.00354975,  0.00752701, -0.00794327, -0.00778868, -0.00164652,
        0.0068918 ,  0.00291673,  0.0021475 , -0.0081624 , -0.00310096,
        0.00414062,  0.00169684,  0.00954767,  0.0094126 , -0.00751523,
        0.00489935,  0.00634651,  0.00778395,  0.00034678, -0.00514487,
        0.00600662,  0.00944785, -0.00905489,  0.00700527,  0.00811886,
        0.00404599,  0.00191028, -0.0073443 ,  0.00705282, -0.00915282,
       -0.00484462, -0.00525144, -0.00837096, -0.00492031,  0.00252984],
      dtype=float32)

In [ ]:
model4['GPS']

array([-0.00010812,  0.00654889, -0.0096795 ,  0.00419711,  0.00881517,
        0.00870045, -0.00085944,  0.00877158, -0.0076904 , -0.00458279,
        0.00690111,  0.00925323,  0.00221632, -0.00508633, -0.00584864,
        0.00062931,  0.00525857,  0.00349085, -0.00545431, -0.00413292,
        0.00664355, -0.00995706, -0.00188874,  0.00823313,  0.00427296,
        0.0067082 , -0.00952195, -0.00061338, -0.00569869, -0.0013225 ,
       -0.00407934,  0.00688981,  0.00871297, -0.00707884, -0.00306569,
        0.00214769, -0.00354203,  0.00349761, -0.00123983, -0.00031144,
        0.00984737, -0.00711029,  0.00292753,  0.00291804, -0.00564377,
       -0.00509669, -0.00781621,  0.00239007,  0.0001292 ,  0.00866508],
      dtype=float32)

In [ ]:
model3.most_similar('text')

[('There’s', 0.2794710099697113),
 ('commands,', 0.26954254508018494),
 ('mission-critical', 0.25699689984321594),
 ('consumer', 0.19439628720283508),
 ('dictation', 0.18717510998249054),
 ('employee', 0.18183262646198273),
 ('another,', 0.16919739544391632),
 ('large', 0.15288642048835754),
 ('simplify', 0.12655995786190033),
 ('time.', 0.12331050634384155)]

In [ ]:
model4.most_similar('NLP')

[('help', 0.25218427181243896),
 ('one', 0.213368222117424),
 ('business', 0.18884296715259552),
 ('dictation', 0.1857631653547287),
 ('rapidly—even', 0.16017085313796997),
 ('computer', 0.15341691672801971),
 ('But', 0.14902864396572113),
 ('summarize', 0.14282067120075226),
 ('commands,', 0.13742339611053467),
 ('digital', 0.13535983860492706)]

In [33]:
word_embedding_avg_collectiveiii = []            #Averaged vector for the entire sentence-skipgram
counter = 0
for doc in values:
    vec = get_mean_vector(model3, doc)
    word_embedding_avg_collectiveiii.append(vec)
    if len(vec) > 0:
      counter += 1
      print('Dimension of vector for each word:',vec.shape)

word_embedding_avgiii = np.mean(word_embedding_avg_collectiveiii, axis=0)     
print() 
print('Total words in corpus:',counter)
print('Dimension of vector for overall corpus:',word_embedding_avgiii.shape)

print('Vector for overall corpus:',word_embedding_avgiii) 

Dimension of vector for each word: (50,)
Dimension of vector for each word: (50,)
Dimension of vector for each word: (50,)
Dimension of vector for each word: (50,)
Dimension of vector for each word: (50,)
Dimension of vector for each word: (50,)
Dimension of vector for each word: (50,)
Dimension of vector for each word: (50,)
Dimension of vector for each word: (50,)
Dimension of vector for each word: (50,)
Dimension of vector for each word: (50,)
Dimension of vector for each word: (50,)
Dimension of vector for each word: (50,)
Dimension of vector for each word: (50,)
Dimension of vector for each word: (50,)
Dimension of vector for each word: (50,)
Dimension of vector for each word: (50,)
Dimension of vector for each word: (50,)
Dimension of vector for each word: (50,)
Dimension of vector for each word: (50,)
Dimension of vector for each word: (50,)
Dimension of vector for each word: (50,)
Dimension of vector for each word: (50,)
Dimension of vector for each word: (50,)
Dimension of vec

In [63]:
word_embedding_avg_collectiveiv = []            #Averaged vector for the entire sentence-skipgram
counter = 0
for doc in values:
    vec = get_mean_vector(model4, doc)
    word_embedding_avg_collectiveiv.append(vec)
    if len(vec) > 0:
      counter += 1
      print('Dimension of vector for each word:',vec.shape)

word_embedding_avgiv = np.mean(word_embedding_avg_collectiveiv, axis=0)     
print() 
print('Total words in corpus:',counter)
print('Dimension of vector for overall corpus:',word_embedding_avgiv.shape)

print('Vector for overall corpus:',word_embedding_avgiv) 

Dimension of vector for each word: (50,)
Dimension of vector for each word: (50,)
Dimension of vector for each word: (50,)
Dimension of vector for each word: (50,)
Dimension of vector for each word: (50,)
Dimension of vector for each word: (50,)
Dimension of vector for each word: (50,)
Dimension of vector for each word: (50,)
Dimension of vector for each word: (50,)
Dimension of vector for each word: (50,)
Dimension of vector for each word: (50,)
Dimension of vector for each word: (50,)
Dimension of vector for each word: (50,)
Dimension of vector for each word: (50,)
Dimension of vector for each word: (50,)
Dimension of vector for each word: (50,)
Dimension of vector for each word: (50,)
Dimension of vector for each word: (50,)
Dimension of vector for each word: (50,)
Dimension of vector for each word: (50,)
Dimension of vector for each word: (50,)
Dimension of vector for each word: (50,)
Dimension of vector for each word: (50,)
Dimension of vector for each word: (50,)
Dimension of vec

In [53]:
#Doc2Vec                        #Averaged vector for the entire sentence DOC2VEC
temp = values
temp2ii = []
for i in temp:
    for j in i:
        temp2ii.append(j)
import gensim
docmodelii = gensim.models.doc2vec.Doc2Vec(vector_size=50, min_count=2, epochs=50)

#building vocab
#docmodel.build_vocab([s.encode('utf-8').split() for s in temp2ii])


sentences = [gensim.models.doc2vec.TaggedDocument(sentence, 'tag') for sentence in temp2ii]
docmodelii.build_vocab(sentences)

In [54]:
docmodelii.train(sentences, total_examples=docmodelii.corpus_count, epochs=docmodelii.epochs)

In [55]:
vectorii = docmodelii.infer_vector(temp2ii)    #dim 50
print(vectorii)                             

[ 0.00088993 -0.00712357  0.00204865 -0.00830369  0.00308902  0.00716895
  0.00359111 -0.00634366 -0.00768017  0.00855012  0.00871506  0.00488005
 -0.00846803 -0.00307154 -0.00663422 -0.00210881  0.00351603 -0.00541055
  0.00253019 -0.00269983  0.00804262  0.00401383  0.00844861  0.00689903
  0.00850768  0.00527708  0.00499489  0.00880536  0.00981555  0.00917694
 -0.00877941  0.00019085 -0.00945838 -0.00265259  0.00590502  0.00819771
  0.0083445   0.00250356 -0.00436055 -0.00926238 -0.00361688 -0.00864589
 -0.00965511 -0.00711285  0.00168902  0.00193385 -0.00681643 -0.00939186
 -0.00882162  0.00536393]


In [66]:
result2 = 1 - spatial.distance.cosine(word_embedding_avgiv, vectorii)   
result2 

0.18631403148174286

#Conclusion:
1) Word embedding is one of the most important techniques in natural language processing(NLP), where words are mapped to vectors of real numbers.        
2) Word embedding is capable of capturing the meaning of a word in a document, semantic and syntactic similarity, relation with other words.       
3) For both task 2&3 we can observe that CBOW & Skipgram feature vector formed for the words are same i.e. there is no difference.              
4) When we check the cosine similarity between the avg w2v and d2v vector for Task 2 we get -0.22 meaning the vectors are dissimilar and have negative direction/correlation.                                         
Whereas in task 3 doing the same on the new vectors, we got a cosine similarity of +0.186 meaning the vectors are similar and have positive correlation.                     
 This suggests that having large enough corpus (which we had for task3) will help to make better models of CBOW/Skipgram/D2V since they all showed some positive correlation amongst themselves meaning they were able to train better on this corpus than the one in task2.               

In [ ]:
! jupyter nbconvert --to html I082_Shreyas_Dongre_NLP_Exp7.ipynb